File: c003/001 - binary trees/js/Tree.js
module.exports = class Tree {
  constructor (value, parent = null) {
    this._parent = parent
    this._value = value
    this._left = null
    this._right = null
  }

  insert (tree) {
    if (tree._value < this._value) {
      if (this._left === null) {
        tree._parent = this
        this._left = tree
      } else {
        this._left.insert(tree)
      }
    } else if (tree._value > this._value) {
      if (this._right === null) {
        tree._parent = this
        this._right = tree
      } else {
        this._right.insert(tree)
      }
    }
  }

  print (order = false) {
    if (order) {
      if (this._right !== null) this._right.print(order)
      console.log(this._value)
      if (this._left !== null) this._left.print(order)
    } else {
      if (this._left !== null) this._left.print(order)
      console.log(this._value)
      if (this._right !== null) this._right.print(order)
    }
  }

  search (value) {
    if (this._value === value) return this
    if (this._left !== null && value < this._value) {
      return this._left.search(value)
    } else if (this._right !== null && value > this._value) {
      return this._right.search(value)
    }
    return null
  }

  max () {
    if (this._right !== null) return this._right.max()
    return this._value
  }

  min () {
    if (this._left !== null) return this._left.min()
    return this._value
  }
  
  getMinNode () {
    if (this._left !== null) return this._left.getMinNode()
    return this

  }

  delete (value, root = this) {
    if (root == null) return null
    if (value < root._value) {
      root._left = root._left.delete(value)
    } else if (value > root._value) {
      root._right = root._right.delete(value)
    } else {
      // case 1: no child
      if (root._left === null && root._right === null) {
        root = null        
      } else if (root._left === null) { // case 2: one child
        let temp = root
        root = root._right
        temp = null
      } else if (root._right === null) {
        let temp = root
        root = root._left
        temp = null
      } else { // case 3: 2 children
        let minNode = root._right.getMinNode()
        root._value = minNode._value
        root._right = root._right.delete(minNode._value)
      }
    }
    return root
  }
}

File: c003/001 - binary trees/py/classes/tree.py
class Tree:

    def __init__(self, value, parent=None):
        self.__value = value
        self.__parent = parent
        self.__left = None
        self.__right = None


    # getters
    def get_value(self):
        return self.__value


    def get_left(self):
        return self.__left


    def get_right(self):
        return self.__right


    # setters
    def set_parent(self, parent):
        self.__parent = parent


    def set_left(self, left):
        self.__left = left


    def set_right(self, right):
        self.__right = right

    def set_value(self, value):
        self.__value = value


    # methods
    def insert(self, tree):
        if tree.get_value() < self.get_value():
            if self.get_left() is None:
                tree.set_parent(self)
                self.set_left(tree)
            else:
                self.get_left().insert(tree)
        elif tree.get_value() > self.get_value():
            if self.get_right() is None:
                tree.set_parent(self)
                self.set_right(tree)
            else:
                self.get_right().insert(tree)


    def print(self, order=False):
        if order:
            if self.get_right() is not None:
                self.get_right().print(order)
            print(self.get_value())
            if self.get_left() is not None:
                self.get_left().print(order)
        else:
            if self.get_left() is not None:
                self.get_left().print(order)
            print(self.get_value())
            if self.get_right() is not None:
                self.get_right().print(order)


    def search(self, value):
        if self.get_value() == value:
            return self

        if self.get_left() is not None and value < self.get_value():
            return self.get_left().search(value)
        elif self.get_right() is not None and value > self.get_value():
            return self.get_right().search(value)

        return None


    def min(self):
        if self.get_left() is not None:
            return self.get_left().min()
        return self


    def max(self):
        if self.get_right() is not None:
            return self.get_right().max()
        return self


    def delete(self, value):
        if self is None:
            return None
        if value < self.get_value() and self.get_left() is not None:
            self.set_left(self.get_left().delete(value))
        elif value > self.get_value() and self.get_right() is not None:
            self.set_right(self.get_right().delete(value))
        else:
            # case 1: no child
            if self.get_left() is None and self.get_right() is None:
                self = None
            elif self.get_left() is None: # case 2: one child
                temp = self
                self = self.get_right()
                temp = None
            elif self.get_right() is None:
                temp = self
                self = self.get_left()
                temp = None
            else: # case 3: 2 children
                print('i am here')
                min_node = self.get_right().min()
                self.set_value(min_node.get_value())
                print('right', self.get_right().get_value())
                temp = self.get_right().delete(min_node.get_value())
                print('test this', temp)
                self.set_right(temp)
        return self

